kubetest/kind.go

```diff
@@ -0,0 +1,177 @@
+	// Ensure the kind binary path is added in $PATH.
+	// Could be redundant if kindBinaryPath is /go/bin and this is already in $PATH.
+	path := os.Getenv("PATH")
+	if !strings.Contains(path, kindBinaryPath) {
+		if err := os.Setenv("PATH", path+":"+kindBinaryPath); err != nil {
+			return nil, err
+		}
+	}
+	k := &kind{
+		configPath: *kindConfigPath,
+	}
+	return k, nil
+}
+
+func getKubeConfigPath() (string, error) {
+	o, err := control.Output(exec.Command("kind", "get", "kubeconfig-path"))
+	if err != nil {
+		return "", err
+	}
+	return strings.TrimSuffix(string(o), "\n"), nil
+}
+
+func setKubeConfigEnv() error {
+	path, err := getKubeConfigPath()
+	if err != nil {
+		return err
+	}
+	if err = os.Setenv("KUBECONFIG", path); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Builds kind an the node image and deploys a cluster based on the kind config.
+func (k *kind) Up() error {
+	// Build the kind binary.
+	cmd := exec.Command("go", "build")
+	cmd.Dir = kindSrcPath
+	if err := control.FinishRunning(cmd); err != nil {
+		return err
+	}
+	// Copy the kind binary to the kind binary path.
+	cmd = exec.Command("cp", "./kind", kindBinaryPath)
+	cmd.Dir = kindSrcPath
+	if err := control.FinishRunning(cmd); err != nil {
+		return err
+	}
+	// Build the node image.
+	cmd = exec.Command("kind", "build", "node-image", "--image="+kindNodeImage)
+	if err := control.FinishRunning(cmd); err != nil {
+		return err
+	}
+	// Build kubectl.
+	cmd = exec.Command("make", "all", "WHAT=cmd/kubectl")
+	cmd.Dir = k8sSrcPath
+	if err := control.FinishRunning(cmd); err != nil {
+		return err
+	}
+	// Copy kubectl to the kind binary path.
+	cmd = exec.Command("cp", "./_output/local/go/bin/kubectl", kindBinaryPath)
+	cmd.Dir = k8sSrcPath
+	if err := control.FinishRunning(cmd); err != nil {
+		return err
+	}
+	// Create the kind cluster.
+	useConfig := ""
+	if k.configPath != "" {
+		useConfig = "--config=" + k.configPath
+	}
+	cmd = exec.Command("kind", "create", "cluster", "--image="+kindNodeImage,
+		"--retain", "--wait=1m", "--loglevel=debug",
+		useConfig)
+	if err := control.FinishRunning(cmd); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (k *kind) IsUp() error {
+	return isUp(k)
+}
+
+func (k *kind) DumpClusterLogs(localPath, gcsPath string) error {
+	cmd := exec.Command("kind", "export", "logs", localPath)
+	if err := control.FinishRunning(cmd); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (k *kind) TestSetup() error {
+	// set KUBECONFIG
+	if err := setKubeConfigEnv(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (k *kind) Down() error {
+	cmd := exec.Command("kind", "delete", "cluster")
+	if err := control.FinishRunning(cmd); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (k *kind) GetClusterCreated(gcpProject string) (time.Time, error) {
+	return time.Time{}, errors.New("not implemented")
+}
+
+func (k *kind) KubectlCommand() (*exec.Cmd, error) {
+	// Avoid using ./cluster/kubectl.sh ?
```

Comment:
yes! let's not depend on these scripts

Reply:
probably we will need to let `--deployer=kind` hook the kubetest `--build` process to instead do the appropriate `--type` of `kind build node-image` and a kubectl build for the host.

Reply:
i need to investigate how `build` works and why it's needed.
will add a TODO.